### PR TITLE
fix: calling `evm_mine` with a `timestamp` argument should reflect the change of time in subsequent blocks

### LINKED ISSUE #3265: calling `evm_mine` with a `timestamp` does not persist time offset
### Discussed in https://github.com/trufflesuite/ganache/discussions/3264

I'm opening as  an issue as suggested by @davidmurdoch

<div type='discussions-op-text'>

<sup>Originally posted by **msuscens** June 21, 2022</sup>
Hello, 

I have Truffle tests that successfully pass with Ganache v6 but fail with Ganache v7.3.1.  The truffle tests in question use the ganache-time-traveler utility to advance time.  Specifically, the issue is with functions that need a certain amount of time to pass before they will execute (without reverting).  I, therefore, use ganache-time-traveler to advance the time before executing the function.  Unfortunately, with ganache 7.3.1 these functions revert even though time has been advanced by the required amount.   Here is an example:

```
describe("After Incubation: Hatch single egg", () => {

        let snapshotId
        beforeEach("Advance Time to End of Incubation", async() => {
            let snapshot = await timeMachine.takeSnapshot()
            snapshotId = snapshot['result']

            // Incubate for required time
            await truffleAssert.passes(
                timeMachine.advanceTimeAndBlock(INCUBATION_DURATION),
                "Failed to advance time and block"
            )
        })

        afterEach("Revert To Start of Incubation Time", async() => {
            await timeMachine.revertToSnapshot(snapshotId)
        })

        it("should allow egg owner to hatch it, emitting an Hatched event ", async () => {
            let txResult
            await truffleAssert.passes(
                txResult = await eggToken.hatch(
                    [0], //eggids
                    {from: accounts[2]}
                )
            )
            truffleAssert.eventEmitted(txResult, 'Hatched', (ev) => {
                return equals(toNumbers(ev.hatchlingIds), [0]) &&
                    equals(toNumbers(ev.eggIds),[0]) &&
                    ev.owner == accounts[2]
            }, "Event Hatched has incorrect parameter values!")
        })
```
etc ...

This test passes with Ganche v6 but `eggToken.hatch` reverts under v7.3.1 with the following message:

```
After Incubation: Hatch single egg
         should allow egg owner to hatch it, emitting an Hatched event :
     Transaction: 0x20883a2a3329b9925a86499b4fe1e9727a029f197d00f084ed920e94eb1954d6 exited with an error (status 0). Reason given: hatch: Egg not incubated!.
     Please check that the transaction:
     - satisfies all conditions set by Solidity `require` statements.
     - does not trigger a Solidity `revert` statement.
 
  StatusError: Transaction: 0x20883a2a3329b9925a86499b4fe1e9727a029f197d00f084ed920e94eb1954d6 exited with an error (status 0). Reason given: hatch: Egg not incubated!.
      at Context.<anonymous> (test/2.eggToken/6.eggToken_hatch_gen0_test.js:217:43)
      at runMicrotasks (<anonymous>)
      at processTicksAndRejections (internal/process/task_queues.js:93:5)
```

I'd appreciate any help/advice on how to resolve this issue.

I'm on a 2022 MacBook Pro (M1 Pro processor): MacOS 12.4
Truffle v5.5.3 (core: 5.5.3)
Ganache v7.0.1
Solidity - 0.8.13 (solc-js)
Node v14.15.5
Web3.js v1.5.3

Note: Even though it says "Ganache v7.0.1" above (from 'truffle version'), I've installed v7.3.1 and when I start ganache (via CL) it says "ganache v7.3.1 (@ganache/cli: 0.4.1, @ganache/core: 0.4.1)"
</div>

### PR DESCRIPTION
**Pull request by jeffsmale90**
Previously, calling `evm_mine` with a `timestamp` argument would result in a block with the specified timestamp, but subsequent blocks would have a timestamp that didn't reflect this change in time. This only occurred when `miner.timestampIncrement` is unspecified or `clock`. 

```javascript
const provider = ganache.provider();

await provider.request({ method: "evm_mine", params: ["0x1"] });
const { timestamp: firstTimestamp } = await provider.request({
  method: "eth_getBlockByNumber",
  params: ["latest", false],
});

// wait 1 second before mining the second block
await new Promise((resolve, reject) => setTimeout(resolve, 1000));

await provider.request({ method: "evm_mine", params: [] });
const { timestamp: secondTimestamp } = await provider.request({
  method: "eth_getBlockByNumber",
  params: ["latest", false],
});

console.log({
  firstTimestamp,
  secondTimestamp,
});
```

Will output something like:
```
{ firstTimestamp: '0x1', secondTimestamp: '0x5e0be0ff' }
```

Where `secondTimestamp` is the current time in seconds, but should be `0x2`.

With this change, blocks mined after providing a `timestamp` parameter to `evm_mine`, will have timestamps that reflect the change in time.

Fixes: #3265

### CODE REVIEW

File: src/chains/ethereum/ethereum/tests/blockchain.test.ts [hunk 1]
```diff
@@ -112,7 +112,12 @@ describe("blockchain", async () => {
       // assert that second block's timestamp is at least `blockTime` greater
       // than the first block's. meaning, these blocks weren't mined one after
       // the other
-      assert(timestamps[1] >= timestamps[0] + blockTime);
+      assert(
+        timestamps[1] >= timestamps[0] + blockTime,
+        `Unexpected timestamp - expected >= ${timestamps[0] + blockTime}, got ${
```
Comment (jeffsmale90) [inline src/chains/ethereum/ethereum/tests/blockchain.test.ts:117]: I updated this message as I was debugging the change, and the error message wasn't helpful

File: src/chains/ethereum/ethereum/src/blockchain.ts [hunk 2]
```diff
@@ -833,7 +841,8 @@ export default class Blockchain extends Emittery<BlockchainTypedEvents> {
   }
 
   /**
-   * @param newTime - the number of milliseconds to adjust the time by. Can be negative.
```
Comment (jeffsmale90) [inline src/chains/ethereum/ethereum/src/blockchain.ts:842]: I found this comment misleading - happy to hear thoughts on whether my change is better or not.

File: src/chains/ethereum/ethereum/tests/provider.test.ts [hunk 3]
```diff
@@ -141,6 +141,84 @@ describe("provider", () => {
...
+
+        assert.strictEqual(
+          specifiedBlock,
+          `0x${timeArgumentSeconds.toString(16)}`,
+          "Unexpected timestamp for block mined with specified timestamp"
+        );
+
+        assert(
+          +unspecifiedBlock >= timeArgumentSeconds &&
+            +unspecifiedBlock <= timeArgumentSeconds + 1000,
```
Comment (MicaiahReid) [inline src/chains/ethereum/ethereum/tests/provider.test.ts]: If CI is running slowly, could this break? I don't know what the solution is, and I don't feel it's worth holding up this PR for, but just wanted to throw it out there in case you have a solution on-hand.

**Review by MicaiahReid — APPROVED** [review]

File: src/chains/ethereum/ethereum/tests/provider.test.ts (hunk 3 above)
Comment (jeffsmale90) [reply src/chains/ethereum/ethereum/tests/provider.test.ts]: It _could_ do. Because we are specifying a timestamp in seconds, it would need to take an entire second to mine and fetch 2 blocks.

I feel like I'm building a lot of pressure on the time refactor - but it'll let us inject a `now` provider, which allow us to be much more resilient to slow build agents.

🤞

File: src/chains/ethereum/ethereum/src/blockchain.ts [hunk 4]
```diff
@@ -360,7 +360,7 @@ export default class Blockchain extends Emittery<BlockchainTypedEvents> {
         const mineAll = (maxTransactions: Capacity, onlyOneBlock = false) =>
           this.#isPaused()
             ? nullResolved
-            : this.mine(maxTransactions, null, onlyOneBlock);
+            : this.mine(maxTransactions, undefined, onlyOneBlock);
```
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/src/blockchain.ts]: Passing `undefined` means we have defined the value as `undefined`, and thus it is defined, with the value `undefined`. I think that is what `null` is for :-) Thoughts?

Also, I think passing `undefined` like this will not be allowed once we enable typescript's "strict" mode (https://github.com/trufflesuite/ganache/issues/3555).

**Review by davidmurdoch — CHANGES_REQUESTED** [review]

File: src/chains/ethereum/ethereum/src/blockchain.ts (hunk 4 above)
Comment (jeffsmale90) [reply src/chains/ethereum/ethereum/src/blockchain.ts]: As discussed, this is probably up for a refinement of the `blockchain` interface. In the meantime, I've reordered the arguments to avoid needing to pass `undefined` or `null` explicitly.

(I also named the arguments within `evm_mine()` in api.ts because I found them confusing - open to dissenting opinions though!)

File: src/chains/ethereum/ethereum/src/blockchain.ts [hunk 5]
```diff
@@ -608,6 +608,15 @@ export default class Blockchain extends Emittery<BlockchainTypedEvents> {
       onlyOneBlock
     );
     await this.#blockBeingSavedPromise;
+
+    if (
+      timestamp !== undefined &&
+      this.#options.miner.timestampIncrement === "clock"
```
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/src/blockchain.ts]: There are two time-related blocks of code in this function that both check `this.#options.miner.timestampIncrement === "clock"`. Can these two blocks be joined together? If not, can you cache `this.#options.miner.timestampIncrement === "clock"` in a variable so we don't do it twice?

File: src/chains/ethereum/ethereum/tests/provider.test.ts [hunk 6]
```diff
@@ -141,6 +141,83 @@ describe("provider", () => {
       await provider.disconnect();
     });
 
+    describe("uses timestamp adjustment in subsequent blocks after calling `evm_mine` with a `timestamp` argument", () => {
```
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/tests/provider.test.ts]: Our `describe`s are usually simple words or phrases like "api", "eth_chainId", "server". This is worded more like an `it`.

Can you rephrase the `describe` and `it`s here to be more inline with our other tests?

We use `describe` to define _structure_ to our tests, so maybe a bunch of these tests in this file should already be under `describe("timestampIncrement", ...)` (don't need to make a change like this in this same PR, of course), and then you could have another `describe` under that one like `describe("`evm_mine` with `timestamp`", ...)` (just an example -- I didn't think it through very much).

File: src/chains/ethereum/ethereum/tests/provider.test.ts [hunk 7]
```diff
@@ -141,6 +141,83 @@ describe("provider", () => {
...
+        await provider.request({
+          method: "evm_mine",
+          params: []
+        });
+        const unspecifiedBlock = await provider.request({
+          method: "eth_getBlockByNumber",
+          params: ["latest", false]
+        });
+
+        return [specifiedBlock.timestamp, unspecifiedBlock.timestamp];
```
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/tests/provider.test.ts]: If you coerce these to numbers here it'd make the tests slightly simpler.

File: src/chains/ethereum/ethereum/tests/provider.test.ts [hunk 8]
```diff
@@ -141,6 +141,83 @@ describe("provider", () => {
...
+            timeArgumentSeconds + 1000
+          }, got ${+unspecifiedBlock}`
+        );
+
+        await provider.disconnect();
+      });
+
+      it("should work with a numeric `timestampIncrement`", async () => {
+        const provider = await getProvider({
+          miner: { timestampIncrement }
```
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/tests/provider.test.ts]: Since this `timestampIncrement` is only used in this function is there a reason to define it in the `describe`?

**Review by davidmurdoch — CHANGES_REQUESTED** [review]

File: src/chains/ethereum/ethereum/src/blockchain.ts (hunk 5 above)
Comment (jeffsmale90) [reply src/chains/ethereum/ethereum/src/blockchain.ts]: Good call!

File: src/chains/ethereum/ethereum/tests/provider.test.ts (hunk 7 above)
Comment (jeffsmale90) [reply src/chains/ethereum/ethereum/tests/provider.test.ts]: I intentionally didn't do this, thinking that we should be asserting the _exact_ returned value. Otherwise, there's a potential issue that could slip through where the return type isn't what's expected, but _coerces to_ what's expected.

Thinking about it in light of your comment, I think that we _should_ be ensuring the above, but maybe in a specific test for that.

I've updated my tests to what you've suggested, keen to hear your thoughts on that.

File: src/chains/ethereum/ethereum/tests/provider.test.ts (hunk 6 above)
Comment (jeffsmale90) [reply src/chains/ethereum/ethereum/tests/provider.test.ts]: I just flattened that describe down, and was more explicit in the test names. lmk what you think.

File: src/chains/ethereum/ethereum/src/blockchain.ts [hunk 9]
```diff
@@ -357,10 +357,10 @@ export default class Blockchain extends Emittery<BlockchainTypedEvents> {
 
         //#region automatic mining
         const nullResolved = Promise.resolve(null);
-        const mineAll = (maxTransactions: Capacity, onlyOneBlock = false) =>
+        const mineAll = (maxTransactions: Capacity, onlyOneBlock?: boolean) =>
```
Comment (davidmurdoch) [inline src/chains/ethereum/ethereum/src/blockchain.ts:360]: A decent alternative would be to change `mineAll` to `(maxTransactions: Capacity, onlyOneBlock: boolean) => `

and then all use of `mineAll` would need to pass the values it needs, basically just:

```javascript
txPool.on("drain", () => mineAll(Capacity.Single, false));
```

This might make this code a bit clearer. Or not. I don't think I care either way.

**Review by davidmurdoch — APPROVED** [review]
Approved! (with one half-hearted suggestion/thought)

File: src/chains/ethereum/ethereum/src/blockchain.ts (hunk 9 above)
Comment (jeffsmale90) [reply src/chains/ethereum/ethereum/src/blockchain.ts:360]: I left it as is because we set a default on the `mine` function, and don't feel that requiring it for `mineAll`, but not for `mine` feels nice - we might as well leverage the default that we have specified.
